test_conformance/images/clReadWriteImage/test_read_1D_buffer.cpp

```diff
@@ -0,0 +1,327 @@
+// limitations under the License.
+//
+#include "../testBase.h"
+#include <CL/cl.h>
+
+int test_read_image_1D_buffer(cl_context context, cl_command_queue queue,
+                              image_descriptor *imageInfo, MTdata d,
+                              cl_mem_flags flags)
+{
+    int error;
+
+    clMemWrapper image;
+    clMemWrapper buffer;
+
+    // Generate some data to test against
+    BufferOwningPtr<char> imageValues;
+    generate_random_image_data(imageInfo, imageValues, d);
+
+    if (gDebugTrace)
+    {
+        log_info(" - Creating %s 1D image %d...\n",
+                 gTestMipmaps ? "mipmapped" : "", (int)imageInfo->width);
+        log_info(" with %llu mip levels\n",
+                 (unsigned long long)imageInfo->num_mip_levels);
+    }
+
+    buffer = clCreateBuffer(context, flags, imageInfo->rowPitch, NULL, &error);
+    if (error != CL_SUCCESS)
+    {
+        log_error("ERROR: Unable to create buffer for 1D image buffer of size "
+                  "%d (%s)",
+                  (int)imageInfo->rowPitch, IGetErrorString(error));
+    }
+
+    // Construct testing sources
+    if (!gTestMipmaps)
+    {
+        image = create_image_1d(context, flags, imageInfo->format,
+                                imageInfo->width, 0, NULL, buffer, &error);
+        if (image == NULL)
+        {
+            log_error("ERROR: Unable to create 1D image buffer of size %d (%s)",
+                      (int)imageInfo->width, IGetErrorString(error));
+            return -1;
+        }
+    }
+    else
+    {
+        cl_image_desc image_desc = { 0 };
+        image_desc.image_type = CL_MEM_OBJECT_IMAGE1D_BUFFER;
+        image_desc.image_width = imageInfo->width;
+        image_desc.num_mip_levels = imageInfo->num_mip_levels;
+        image_desc.buffer = buffer;
+
+        image = clCreateImage(context, flags, imageInfo->format, &image_desc,
+                              NULL, &error);
+        if (error != CL_SUCCESS)
+        {
+            log_error(
+                "ERROR: Unable to create %d level mipmapped 1D image buffer of "
+                "size %d x %d (pitch %d ) (%s)",
+                (int)imageInfo->num_mip_levels, (int)imageInfo->width,
+                (int)imageInfo->height, (int)imageInfo->rowPitch,
+                IGetErrorString(error));
+            return error;
+        }
+    }
+
+    if (gDebugTrace) log_info(" - Writing image...\n");
+
+    size_t origin[3] = { 0, 0, 0 };
+    size_t region[3] = { 0, 1, 1 };
+    size_t fullImageSize;
+    if (gTestMipmaps)
+    {
+        fullImageSize = (size_t)compute_mipmapped_image_size(*imageInfo);
+    }
+    else
+    {
+        fullImageSize = imageInfo->rowPitch;
+    }
+
+    BufferOwningPtr<char> resultValues(malloc(fullImageSize));
+    size_t imgValMipLevelOffset = 0;
+
+    for (size_t lod = 0; (gTestMipmaps && lod < imageInfo->num_mip_levels)
+         || (!gTestMipmaps && lod < 1);
+         lod++)
+    {
+        origin[1] = lod;
```

Comment:
This should be 0 for 1D image or 1D image buffer objects.  

" If image is a 1D image or 1D image buffer object, origin[1] and origin[2] must be 0."

https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueReadImage

Reply:
I don't know much about `gTestMipmaps` and `lod`. This is coming from `test_read_1D.cpp`, which is breaking the specification as you pointed out. Should I just get rid of the for loop and replace all `lod` with `0`?

Reply:
From section 29 of the extension spec, 1D image buffers don't support mipmaps. So we should remove this testing entirely for this object type:
https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_mipmap_image

"Mip-mapped images cannot be created for CL_MEM_OBJECT_IMAGE1D_BUFFER images, depth images or multi-sampled (i.e. msaa) images."

Reply:
Done